File: src/backend/utils/web.py
from fastapi import FastAPI, APIRouter, Request
from fastapi.responses import HTMLResponse, JSONResponse
from fastapi.templating import Jinja2Templates
from fastapi.staticfiles import StaticFiles
from fastapi.middleware.cors import CORSMiddleware
from .logger import logger
from .cost import cost_tracker

# Initialize FastAPI app
app = FastAPI(
    title="Army-GPT",
    description="Collection of AI tools and agents for army dudes",
    version="1.0.0"
)

# Setup templates
templates = Jinja2Templates(directory="src/frontend/templates")

# Setup static files
app.mount("/static", StaticFiles(directory="src/frontend/static"), name="static")

# Setup CORS
app.add_middleware(
    CORSMiddleware,
    allow_origins=["*"],  # In production, replace with specific origins
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
)

# Create router for web pages
router = APIRouter(tags=["web"])

# Create router for API endpoints
api_router = APIRouter(prefix="/api", tags=["api"])

@api_router.get("/costs")
async def get_costs():
    """Get current costs for API usage and server rent"""
    costs = cost_tracker.get_current_costs()
    logger.debug(f"Returning costs: {costs}")  # Debug log
    return JSONResponse(content=costs)

@router.get("/", response_class=HTMLResponse)
async def root(request: Request):
    """Serve the main landing page"""
    return templates.TemplateResponse("index.html", {"request": request})

@router.get("/pace-notes", response_class=HTMLResponse)
async def pace_notes_page(request: Request):
    """Serve the pace notes page"""
    return templates.TemplateResponse("pace-notes.html", {"request": request})

@router.get("/policy-foo", response_class=HTMLResponse)
async def policy_foo_page(request: Request):
    """Serve the policy foo page"""
    return templates.TemplateResponse("policy-foo.html", {"request": request})

# Include both routers
app.include_router(router)
app.include_router(api_router) 

File: src/backend/utils/web.ts
import { FastifyInstance, FastifyRequest, FastifyReply } from 'fastify';
import { join } from 'path';
import { fileURLToPath } from 'url';
import fastifyStatic from '@fastify/static';
import fastifyView from '@fastify/view';
import fastifyCors from '@fastify/cors';
import ejs from 'ejs';
import { config } from '../config';
import { logger } from './logger';

const __filename = fileURLToPath(import.meta.url);
const __dirname = join(__filename, '..');

// Setup web routes
export async function setupWebRoutes(app: FastifyInstance) {
  // Register static file serving
  await app.register(fastifyStatic, {
    root: join(__dirname, '../../frontend/static'),
    prefix: '/static/'
  });

  // Register template engine (using @fastify/view with ejs)
  await app.register(fastifyView, {
    engine: {
      ejs: ejs
    },
    root: join(__dirname, '../../frontend/templates')
  });

  // Web routes
  app.get('/', async (request: FastifyRequest, reply: FastifyReply) => {
    try {
      return reply.view('index.ejs', { request });
    } catch (error) {
      logger.error('Error serving index page:', error);
      return reply.status(500).send('Internal Server Error');
    }
  });

  app.get('/pace-notes', async (request: FastifyRequest, reply: FastifyReply) => {
    try {
      return reply.view('pace-notes.ejs', { request });
    } catch (error) {
      logger.error('Error serving pace notes page:', error);
      return reply.status(500).send('Internal Server Error');
    }
  });

  app.get('/policy-foo', async (request: FastifyRequest, reply: FastifyReply) => {
    try {
      return reply.view('policy-foo.ejs', { request });
    } catch (error) {
      logger.error('Error serving policy foo page:', error);
      return reply.status(500).send('Internal Server Error');
    }
  });

  // API routes
  app.get('/api/costs', async (request: FastifyRequest, reply: FastifyReply) => {
    try {
      // TODO: Implement cost tracking
      const costs = { api: 0, server: 0 }; // Placeholder
      logger.debug('Returning costs:', costs);
      return reply.send(costs);
    } catch (error) {
      logger.error('Error getting costs:', error);
      return reply.status(500).send({
        error: 'Internal server error'
      });
    }
  });

  // Register CORS
  await app.register(fastifyCors, {
    origin: true, // In production, replace with specific origins
    methods: ['GET', 'POST', 'PUT', 'DELETE', 'OPTIONS'],
    allowedHeaders: ['Content-Type', 'Authorization', 'X-Request-ID'],
    exposedHeaders: ['X-RateLimit-Remaining-Hour', 'X-RateLimit-Remaining-Day'],
    credentials: true
  });

  // Error handler
  app.setErrorHandler((error, request, reply) => {
    logger.error('Unhandled error:', error);
    reply.status(500).send({
      error: 'Internal Server Error'
    });
  });
} 